### ISSUE TITLE
Debuginfo tests don't control for ambient tools

### ISSUE THREAD

**Issue by workingjubilee** [rust-lang/rust#126092]
The debuginfo tests should probably use a specific debugger, like maybe an lldb we build with the compiler, so that they work consistently. In practice, they have [been effectively broken on user's machines](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/How.20do.20I.20skip.20debuginfo.20tests.3F) by just trying to use whatever is installed.

In other words: just because I use Linux doesn't mean I have gdb installed.

**Comment by workingjubilee** [comment]
The debuginfo test suite should also probably be better insulated against *gdb changing how it prints things, **Because It Does***.

**Comment by Nilstrieb** [comment]
Some points:
- compiletest supports `cdb`, `gdb` and `lldb`, running the test suite for all the ones that are available (either via `config.toml` or via `PATH`).
- our LLVM builds do currently not contain lldb.

**Comment by workingjubilee** [comment]
That's... informative, thank you!

It does seem like we should reassess our strategy if we're gonna run a test suite that isn't guaranteed to pass even between different versions of a debugger on "any ol' random debugger lying around".

**Comment by bjorn3** [comment]
On macOS we can't ship our own debugger. Only an Apple signed lldb is allowed to use the debugging apis of the kernel.

**Comment by workingjubilee** [comment]
Wait, gdb is useless on macOS?

**Comment by bjorn3** [comment]
Supposedly you can use gdb on macOS jf you sign it with a self-signed code signing certificate which you add to the system keychain, mark as trusted for code signing and set the right entitlements when signing gdb. I don't know if you need to disable SIP for this to work though. We can't disable SIP on CI. https://dev.to/jasonelwood/setup-gdb-on-macos-in-2020-489k

**Comment by workingjubilee** [comment]
Hmm, okay. I think on Linux it is still desirable to control for the sheer amount of variability that Linux distros have, but I can believe that the debuginfo tests have to draw on the environment sometimes.

**Comment by jieyouxu** [comment]
I think it's fine if we can't necessarily control for the debugger used, but we should at least document it and/or report the debugger used when running the debuginfo test suite.

**Comment by workingjubilee** [comment]
yeah, that was also reported as https://github.com/rust-lang/rust/issues/126628

**Comment by saethlin** [comment]
FYI all of the lldb tests have been failing for anyone who uses a rolling distro for the past 8 months because of https://github.com/llvm/llvm-project/issues/70453

**Comment by jieyouxu** [comment]
In case anyone stumbled across this, IIRC lldb debuginfo tests (currently) require Python 3.10 dll to be present in `PATH` on Windows, not sure about Unixes but presumably only working with Python 3.10 is well-behaved.

**Comment by jieyouxu** [comment]
this has come back to bite us in https://github.com/rust-lang/rust/issues/133107

**Comment by jyn514** [comment]
i think we should consider the version of the debugger to be part of the "prerequisites" compiletest uses to determine whether to run the test at all (see https://github.com/rust-lang/rust/issues/126628). it doesn't make any sense to run a test that needs lldb 19 against lldb 12.

**Comment by jieyouxu** [comment]
> i think we should consider the version of the debugger to be part of the "prerequisites" compiletest uses to determine whether to run the test at all (see [#126628](https://github.com/rust-lang/rust/issues/126628)). it doesn't make any sense to run a test that needs lldb 19 against lldb 12.

I think there are two parts to this:

1. I think it's reasonable if they are conditioned on prerequisites being available for them to be eligible to run (by default).
2. But there needs to be a way (CI should force, locally should opt-in) to ensure debuginfo tests do get run and not trivially ignored.

**Comment by clarfonthey** [comment]
Just want to comment again on this that these tests constantly get in the way of me locally testing other tests that are failing, since the bootstrap runner will stop running further tests once it gets to the failing debuginfo tests.

Perhaps in the meantime, whenever the debuginfo tests fail, they could output a comment linking to this issue and mentioning that adding `--skip tests/debuginfo` will remove them from the test suite? That way people running tests are reminded quickly of the current status and how to avoid them while not affecting whether they get run in CI at all.
